### PR TITLE
Unify local and CI build scripts with build.ps1

### PR DESCRIPTION
**Pull request by tristanlabelle**
Introduces `build.ps1`, a PowerShell-based build script which is suitable for local incremental builds as well as CI builds, and updates the `build-windows-toolchain.bat` to call it. This will simplify reproducing and investigating build breaks. https://github.com/compnerd/swift-build has been using `build.ps1` for a few weeks now for local builds while we were iterating on it.

`build.ps1` supports building the runtime and SDK for all three architectures, which will help catch regressions. It can be used on amd64 and arm64 hosts. It also minimizes duplication of build defines by factoring common code. For local development, it also supports staging as `S:\Program Files\Swift` and `S:\Library` which can be used like an installed toolchain on the C drive.

### CODE REVIEW

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Comment by compnerd** [comment]
@swift-ci please build toolchain Windows platform

File: utils/build.ps1 [hunk 1]
```diff
@@ -0,0 +1,1323 @@
...
+
+    # Swift Requries COMDAT folding and de-duplication
+    $SwiftArgs.Add("-Xlinker /OPT:REF") | Out-Null
+    $SwiftArgs.Add("-Xlinker /OPT:ICF") | Out-Null
+
+    $SwiftcFlags = $SwiftArgs.ToArray() -Join " "
+    Append-FlagsDefine $Defines CMAKE_Swift_FLAGS $SwiftcFlags
+
+    # Workaround CMake 3.26+ enabling `-wmo` by default on release builds
+    Append-FlagsDefine $Defines CMAKE_Swift_FLAGS_RELEASE "-O"
```
Comment (stevapple) [inline utils/build.ps1:384]: We don’t need this anymore since CMake 3.26.2.

**Review by stevapple — COMMENTED** [review]
This looks good overall. I wonder if we can maintain a mapping from project name to build directory, so we can expressly out the dependency `_DIR`s simply by a list of dependencies. This should reduce the work needed for inserting new dependencies to the workflow, preventing potential mistakes.

**Comment by compnerd** [comment]
> This looks good overall. I wonder if we can maintain a mapping from project name to build directory, so we can expressly out the dependency `_DIR`s simply by a list of dependencies. This should reduce the work needed for inserting new dependencies to the workflow, preventing potential mistakes.

I do like the idea of having a mapping from the project to the directory for the toolchain.  For the SDKs there is the base offset that needs to be accounted for.  Maybe having a toolchain offset (`0`) and the SDK offset makes sense?

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Comment by compnerd** [comment]
@swift-ci please build toolchain Windows platform

**Comment by compnerd** [comment]
@swift-ci please build toolchain Windows platform

**Comment by compnerd** [comment]
@swift-ci please build toolchain Windows platform

**Comment by compnerd** [comment]
@swift-ci please build toolchain Windows platform

**Comment by compnerd** [comment]
@swift-ci please build toolchain Windows platform

**Comment by compnerd** [comment]
```
FAILED: tools/swift/tools/SourceKit/lib/Support/CMakeFiles/SourceKitSupport.dir/Logging.cpp.obj 
T:\1\.\bin\clang-cl.exe  /nologo /TP -DGTEST_HAS_RTTI=0 -DLLVM_ON_WIN32 -DSWIFT_THREADING_WIN32 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_CRT_USE_BUILTIN_OFFSETOF -D_CRT_USE_WINAPI_FAMILY_DESKTOP_APP -D_DLL -D_ENABLE_ATOMIC_ALIGNMENT_FIX -D_ENABLE_EXTENDED_ALIGNED_STORAGE=1 -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_HAS_STATIC_RTTI=0 -D_LIBCPP_ENABLE_ASSERTIONS -D_MD -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools\swift\tools\SourceKit\lib\Support -IC:\Users\swift-ci\jenkins\workspace\swift-PR-build-toolchain-windows\swift\tools\SourceKit\lib\Support -Itools\swift\tools\SourceKit\include -IC:\Users\swift-ci\jenkins\workspace\swift-PR-build-toolchain-windows\swift\tools\SourceKit\include -Itools\swift\include -IC:\Users\swift-ci\jenkins\workspace\swift-PR-build-toolchain-windows\swift\stdlib\public\SwiftShims -IC:\Users\swift-ci\jenkins\workspace\swift-PR-build-toolchain-windows\swift\include -Iinclude -IC:\Users\swift-ci\jenkins\workspace\swift-PR-build-toolchain-windows\llvm-project\llvm\include -Itools\clang\include -IC:\Users\swift-ci\jenkins\workspace\swift-PR-build-toolchain-windows\llvm-project\llvm\..\clang\include -IC:\Users\swift-ci\jenkins\workspace\swift-PR-build-toolchain-windows\cmark\src\include -Itools\cmark\src -IC:\Users\swift-ci\jenkins\workspace\swift-PR-build-toolchain-windows\swift-corelibs-libdispatch\src\BlocksRuntime -IC:\Users\swift-ci\jenkins\workspace\swift-PR-build-toolchain-windows\swift-corelibs-libdispatch -I"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\\include" -I"C:\Program Files (x86)\Windows Kits\10\\Include\10.0.17763.0\ucrt" -I"C:\Program Files (x86)\Windows Kits\10\\Include\10.0.17763.0\shared" -I"C:\Program Files (x86)\Windows Kits\10\\Include\10.0.17763.0\um" /GS- /Gw /Gy /Oi /Oy /Zi /Zc:inline /Zc:__cplusplus /Zc:inline /Zc:__cplusplus /Zi /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd4324 -w14062 -we4238 /Gw /we4062 /wd4068 /permissive- -DOBJC_OLD_DISPATCH_PROTOTYPES=0 /MD /O2 /Ob2  /EHs-c- /GR- -UNDEBUG --target=x86_64-unknown-windows-msvc -O2 /GR- -U_DEBUG -fblocks -std:c++17 /showIncludes /Fotools\swift\tools\SourceKit\lib\Support\CMakeFiles\SourceKitSupport.dir\Logging.cpp.obj /Fdtools\swift\tools\SourceKit\lib\Support\CMakeFiles\SourceKitSupport.dir\SourceKitSupport.pdb /FS -c C:\Users\swift-ci\jenkins\workspace\swift-PR-build-toolchain-windows\swift\tools\SourceKit\lib\Support\Logging.cpp
CreateProcess failed: The system cannot find the file specified.
ninja: fatal: ReadFile: The handle is invalid.
```

This is likely a missing dependency on `clang-cl` for the swapped out compiler.

**Comment by compnerd** [comment]
@swift-ci please build toolchain Windows platform

**Comment by compnerd** [comment]
@swift-ci please build toolchain Windows platform

**Pull request closed without merging**